io.openems.edge.evcs.spelsberg.smart/src/io/openems/edge/evcs/spelsberg/smart/EvcsSpelsbergSmartImpl.java

```diff
@@ -0,0 +1,323 @@
+import io.openems.edge.bridge.modbus.api.AbstractOpenemsModbusComponent;
+import io.openems.edge.bridge.modbus.api.BridgeModbus;
+import io.openems.edge.bridge.modbus.api.ModbusComponent;
+import io.openems.edge.bridge.modbus.api.ModbusProtocol;
+import io.openems.edge.bridge.modbus.api.element.DummyRegisterElement;
+import io.openems.edge.bridge.modbus.api.element.UnsignedDoublewordElement;
+import io.openems.edge.bridge.modbus.api.element.UnsignedWordElement;
+import io.openems.edge.bridge.modbus.api.task.FC16WriteRegistersTask;
+import io.openems.edge.bridge.modbus.api.task.FC3ReadRegistersTask;
+import io.openems.edge.bridge.modbus.api.task.FC6WriteRegisterTask;
+import io.openems.edge.common.channel.value.Value;
+import io.openems.edge.common.component.OpenemsComponent;
+import io.openems.edge.common.event.EdgeEventConstants;
+import io.openems.edge.common.taskmanager.Priority;
+import io.openems.edge.evcs.api.ChargeStateHandler;
+import io.openems.edge.evcs.api.ChargingType;
+import io.openems.edge.evcs.api.Evcs;
+import io.openems.edge.evcs.api.EvcsPower;
+import io.openems.edge.evcs.api.ManagedEvcs;
+import io.openems.edge.evcs.api.Phases;
+import io.openems.edge.evcs.api.Status;
+import io.openems.edge.evcs.api.WriteHandler;
+
+@Designate(ocd = Config.class, factory = true)
+@Component(name = "EVCS.Spelsberg.SMART", immediate = true, configurationPolicy = ConfigurationPolicy.REQUIRE)
+@EventTopics({ EdgeEventConstants.TOPIC_CYCLE_EXECUTE_WRITE })
+public class EvcsSpelsbergSmartImpl extends AbstractOpenemsModbusComponent
+		implements EvcsSpelsbergSmart, Evcs, ManagedEvcs, ModbusComponent, OpenemsComponent, EventHandler {
+
+	@Reference
+	private ConfigurationAdmin cm;
+
+	@Reference
+	private EvcsPower evcsPower;
+
+	@Reference(policy = ReferencePolicy.STATIC, policyOption = ReferencePolicyOption.GREEDY, cardinality = ReferenceCardinality.MANDATORY)
+	protected void setModbus(BridgeModbus modbus) {
+		super.setModbus(modbus);
+	}
+
+	private Config config = null;
+
+	private final Logger log = LoggerFactory.getLogger(EvcsSpelsbergSmartImpl.class);
+
+	private final ChargeStateHandler chargeStateHandler = new ChargeStateHandler(this);
+
+	private final WriteHandler writeHandler = new WriteHandler(this);
+
+	public EvcsSpelsbergSmartImpl() {
+		super(OpenemsComponent.ChannelId.values(), ModbusComponent.ChannelId.values(), Evcs.ChannelId.values(), //
+				ManagedEvcs.ChannelId.values(), EvcsSpelsbergSmart.ChannelId.values());
+	}
+
+	@Activate
+	private void activate(ComponentContext context, Config config) throws OpenemsNamedException {
+		if (super.activate(context, config.id(), config.alias(), config.enabled(), config.modbusUnitId(), this.cm,
+				"Modbus", config.modbus_id())) {
+			return;
+		}
+
+		/*
+		 * Calculates the maximum and minimum hardware power dynamically by listening on
+		 * the fixed hardware limits used for charging
+		 */
+		Evcs.addCalculatePowerLimitListeners(this);
```

Comment:
According to the other discussion about the phase switching, i guess that this static method has to be changed or handled separately, as this method is basically calculating the limits that are shown and used for the ui and final limitations.

Assume a EVCS with hard limits 6A minimum and 32A maximum (In general, power values are preferred in OpenEMS,  it would be more correct to store the hardware limits as current values but for now the fixedMinimumHardware limit is set as a default assuming a 3phase charge).
https://github.com/OpenEMS/openems/blob/ee11345b75053167dd94c541242a37d39a17a69b/io.openems.edge.evcs.api/src/io/openems/edge/evcs/api/Evcs.java#L117
If DC & AC Evcs would be handled different, it would be easier.

fixedMinimumHardwareLimit would be 6A or 4140W, the minimumHardwarePower would be 1380, 2760 or 4140W depending on the current used phases. 
https://github.com/OpenEMS/openems/blob/ee11345b75053167dd94c541242a37d39a17a69b/io.openems.edge.evcs.api/src/io/openems/edge/evcs/api/Evcs.java#L165

For now, you could remove the PowerLimitListeners or adjust it and set the MaximumHardwarePower/MinimumHardwarePower by your own  (Value is the possible amount of power the user and controller can use at this second). For you it would be always the same value for minimum and maximum as you could provide power from 1380 - 11040W.

Try:
 
```
evcs.getMaximumHardwarePowerChannel().setNextValue(11040);
evcs.getMinimumHardwarePowerChannel().setNextValue(1380 );
```
instead of Evcs.addCalculatePowerLimitListeners(this);

Reply:
Okay got it, thanks for the detailed information and implementation hints!
I could replace the PowerLimitListeners by defining static values minimum and maximum power.

However i think there could still be a problem.
What happens if the user will not activate the "automatic phase switching" feature or has an older firmware, which does not support this. Currently i cannot say if is will be possible to determine the activation via a modbus register (i will ask for that). That way, we could react dynamically in a listener and set the limits accordingly.

As for now we have to be compatible with a system, which does not support the full power range of 1380-11040.
Will there be any issues, if we change to static power limits in that case?

Do you see a way to be compatible with both szenarios?

---
We would really like to support our new feature with the Fenecon system as it brings a huge improvement in solar charging, which is a key goal of our product in alignment with Fenecon and Enerix.

Reply:
I think the best way to test if the behaviour is correct would be to test it in a real situation.
In particular the situation where a charge has been completed with a 3-phase charge and a new charge is starting with a 1-phase charge. 

Handling both szenarios could be achieved by getting the information by the configuration (Later AppCenter, so the customer could select) and/or having separate logic depending on the firmware.

As you mentioned: For the first tests (beta) I would not block it because of this feature. I knew a few more things might come up if it was actually used somewhere.